langchain/src/chains/violation_of_expectations/violation_of_expectations_chain.ts

```diff
@@ -0,0 +1,473 @@
+
+/**
+ * Interface for the input parameters of the ViolationOfExpectationsChain class.
+ */
+export interface ViolationOfExpectationsChainInput extends ChainInputs {
+  /**
+   * The retriever to use for retrieving stored
+   * thoughts and insights.
+   */
+  retriever: BaseRetriever;
+  /**
+   * The LLM to use
+   */
+  llm: ChatOpenAI;
+}
+
+/**
+ * Chain that generates key insights/facts of a user based on a
+ * a chat conversation with an AI.
+ */
+export class ViolationOfExpectationsChain
+  extends BaseChain
+  implements ViolationOfExpectationsChainInput
+{
+  static lc_name() {
+    return "ViolationOfExpectationsChain";
+  }
+
+  _chainType(): string {
+    return "violation_of_expectation_chain";
+  }
+
+  chatHistoryKey = "chat_history";
+
+  thoughtsKey = "thoughts";
+
+  get inputKeys() {
+    return [this.chatHistoryKey];
+  }
+
+  get outputKeys() {
+    return [this.thoughtsKey];
+  }
+
+  retriever: BaseRetriever;
+
+  llm: ChatOpenAI;
+
+  jsonOutputParser: JsonOutputFunctionsParser;
+
+  stringOutputParser: StringOutputParser;
+
+  constructor(fields: ViolationOfExpectationsChainInput) {
+    super(fields);
+    this.retriever = fields.retriever;
+    this.llm = fields.llm;
+    this.jsonOutputParser = new JsonOutputFunctionsParser();
+    this.stringOutputParser = new StringOutputParser();
+  }
+
+  getChatHistoryString(chatHistory: BaseMessage[]): string {
+    return chatHistory
+      .map((chatMessage) => {
+        if (chatMessage._getType() === "human") {
+          return `Human: ${chatMessage.content}`;
+        } else if (chatMessage._getType() === "ai") {
+          return `AI: ${chatMessage.content}`;
+        } else {
+          return `${chatMessage.content}`;
+        }
+      })
+      .join("\n");
+  }
+
+  removeDuplicateStrings(strings: Array<string>): Array<string> {
+    return [...new Set(strings)];
+  }
+
+  /**
+   * This method breaks down the chat history into chunks of messages.
+   * Each chunk consists of a sequence of messages ending with an AI message and the subsequent user response, if any.
+   *
+   * @param {BaseMessage[]} chatHistory - The chat history to be chunked.
+   *
+   * @returns {MessageChunkResult[]} An array of message chunks. Each chunk includes a sequence of messages and the subsequent user response.
+   *
+   * @description
+   * The method iterates over the chat history and pushes each message into a temporary array.
+   * When it encounters an AI message, it checks for a subsequent user message.
+   * If a user message is found, it is considered as the user response to the AI message.
+   * If no user message is found after the AI message, the user response is undefined.
+   * The method then pushes the chunk (sequence of messages and user response) into the result array.
+   * This process continues until all messages in the chat history have been processed.
+   */
+  getMessageChunks(chatHistory: BaseMessage[]): MessageChunkResult[] {
```

Comment:
Naming is a vague to me - basically we're returning the conversation to date and whatever the user responded with?

Reply:
Agree my naming on methods could be improved, will fix

Reply:
Refactored to `chunkMessagesByAIResponse`